modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -187,7 +198,18 @@ def row_lengths(self):
         if self._row_lengths_cache is None:
             if len(self._partitions.T) > 0:
                 row_parts = self._partitions.T[0]
-                self._row_lengths_cache = [part.length() for part in row_parts]
+                self._row_lengths_cache = [
+                    part.length(materialize=self._materialize_in_loop)
+                    for part in row_parts
+                ]
+                if not self._materialize_in_loop:
+                    filter_condition = (
+                        self._partition_mgr_cls._execution_wrapper.check_is_future
+                    )
+                    apply_function_on_selected_items(
+                        self._row_lengths_cache, filter_condition, self.materialize_func
+                    )
```

Comment:
wouldn't it be easier to make a static method in `PandasPartition` class that would take a list of partitions and return the lengths in a most efficient way?
```python
class RayPandasPartition(...):
    ...
    @classmethod
    def get_lengths_glob(cls, parts):
        futures = [part._lengths_cache for part in parts]
	return cls._engine_wrapper.materialize(futures)
```
then in our dataframe's code, whenever we want to retrieve lengths/widths for several partitions we would use this method instead:
```suggestion
                self._row_lengths_cache = row_parts[0].get_lengths_glob(row_parts)
```

Reply:
An issue I feel with this new approach is if _lengths_cache would be none, as we have seemingly slightly different logic to compute length cache for each backend. The logic present in .length() function which would have to be replicated in all PandasPartition class implementations.
Also calling the materialize function directly (without `apply_function_on_selected_items`) would fail if some of these values are already materialized. Or I could call lengths directly in the static class and do as below

`class RayPandasPartition(...):
    ...
    @classmethod
    def get_lengths_glob(cls, parts):
        _row_lengths_cache  =  [
                    part.length(materialize=self._materialize_in_loop)
                    for part in row_parts
                ]
	filter_condition = (
                        cls._engine_wrapper.check_is_future
                    )
    apply_function_on_selected_items(
                        self._row_lengths_cache, filter_condition, cls._engine_wrapper.materialize_func
                    )
	return cls._engine_wrapper.materialize(futures)`

Another alternative approach would be to move properties `row_lengths` and `column_widths` to `class PandasOnRayDataframe`. I think Moving concat `class PandasOnRayDataframe`  will cause a lot of redundancy which is why I kept everything in PandasDataframe class itself. eg 

```
class PandasOnRayDataframe(PandasDataframe):
	@property
    def column_widths(self):
        """
        Compute the column partitions widths if they are not cached.

        Returns
        -------
        list
            A list of column partitions widths.
        """
        if self._column_widths_cache is None:
            if len(self._partitions) > 0:
                col_parts = self._partitions[0]

                self._column_widths_cache = [
                    part.width(materialize=self._materialize_in_loop)
                    for part in col_parts
                ]
        
            filter_condition = (
                self._partition_mgr_cls._execution_wrapper.check_is_future
            )
            apply_function_on_selected_items(
                self._column_widths_cache,
                filter_condition,
                self.materialize_func,
            )
            else:
                self._column_widths_cache = []
        return self._column_widths_cach
```

Which of this would be a better approach?

Reply:
the first approach looks better to me

Reply:
Tried out the first approach, however this would require having get_lengths_glob function repeated in block and virtual partitions, and for the implementation specific changes(eg ray) also would have to be replicated on PandasOnRayDataframeVirtualPartition and PandasOnRayDataframePartition, so I think will switch to the second option I mentioned in the comment.